File: packages/unity-csharp/Tests/Runtime/DeepLinkHandlerTests.cs
using System.Threading.Tasks;
using NUnit.Framework;
using Cinacoin;

namespace Cinacoin.Tests.Runtime
{
    /// <summary>
    /// Tests for DeepLinkHandler: URI parsing, wallet routing.
    /// </summary>
    public class DeepLinkHandlerTests
    {
        private DeepLinkHandler _handler;

        [SetUp]
        public void Setup()
        {
            _handler = new DeepLinkHandler();
        }

        // ─── URI Parsing ─────────────────────────────────────────────────

        [Test]
        public void GenerateDeepLink_ForMetaMask()
        {
            var link = _handler.GenerateDeepLink("metamask", "wc:abc123");
            Assert.That(link, Does.StartWith("metamask://"));
            Assert.That(link, Does.Contain("wc:abc123"));
        }

        [Test]
        public void GenerateDeepLink_ForPhantom()
        {
            var link = _handler.GenerateDeepLink("phantom", "wc:sol123");
            Assert.That(link, Does.StartWith("phantom://"));
        }

        [Test]
        public void GenerateDeepLink_ForUnknownWallet_Throws()
        {
            Assert.Throws<System.ArgumentException>(
                () => _handler.GenerateDeepLink("unknown_wallet", "wc:test")
            );
        }

        [Test]
        public void GenerateDeepLink_ForWalletConnect()
        {
            var link = _handler.GenerateDeepLink("walletconnect", "wc:relay123");
            Assert.That(link, Does.StartWith("wc://"));
        }

        // ─── Wallet Detection ────────────────────────────────────────────

        [Test]
        public void GenerateUniversalLink_ForMetaMask()
        {
            var link = _handler.GenerateUniversalLink("metamask", "wc:abc123");
            Assert.That(link, Does.Contain("metamask.app.link"));
        }

        [Test]
        public void GenerateUniversalLink_ForUnknownWallet_Throws()
        {
            Assert.Throws<System.ArgumentException>(
                () => _handler.GenerateUniversalLink("unknown_wallet", "wc:test")
            );
        }

        [Test]
        public void WalletRegistry_HasExpectedWallets()
        {
            Assert.IsTrue(WalletRegistry.Has("metamask"));
            Assert.IsTrue(WalletRegistry.Has("walletconnect"));
            Assert.IsTrue(WalletRegistry.Has("rainbow"));
            Assert.IsTrue(WalletRegistry.Has("phantom"));
        }

        [Test]
        public void WalletRegistry_GetAll_ReturnsNonEmpty()
        {
            var all = WalletRegistry.GetAll();
            Assert.IsNotEmpty(all);
        }
    }
}

File: packages/ios-swift/Tests/OnChainUXTests/DeepLinkHandlerTests.swift
import XCTest
@testable import Cinacoin

final class DeepLinkHandlerTests: XCTestCase {

    // MARK: - URI Parsing

    func testGenerateDeepLinkForMetamask() {
        let handler = DeepLinkHandler()
        let link = handler.generateDeepLink(walletId: "metamask", uri: "wc:abc123")
        XCTAssertTrue(link.hasPrefix("metamask://"))
        XCTAssertTrue(link.contains("wc:abc123"))
    }

    func testGenerateDeepLinkForPhantom() {
        let handler = DeepLinkHandler()
        let link = handler.generateDeepLink(walletId: "phantom", uri: "wc:xyz789")
        XCTAssertTrue(link.hasPrefix("phantom://"))
        XCTAssertTrue(link.contains("wc:xyz789"))
    }

    // MARK: - Wallet Detection

    func testKnownWalletsGenerateLinks() {
        let handler = DeepLinkHandler()
        let wallets = ["metamask", "walletconnect", "coinbase", "rainbow", "trust", "phantom", "rabby"]
        for walletId in wallets {
            let link = handler.generateDeepLink(walletId: walletId, uri: "wc:test")
            XCTAssertFalse(link.isEmpty, "Expected link for \(walletId)")
        }
    }

    func testUnknownWalletReturnsEmpty() {
        let handler = DeepLinkHandler()
        let link = handler.generateDeepLink(walletId: "unknown_wallet", uri: "wc:12345")
        XCTAssertEqual(link, "")
    }

    func testGenerateDeepLinkWithQueryParams() {
        let handler = DeepLinkHandler()
        let link = handler.generateDeepLink(
            walletId: "metamask",
            uri: "wc:abc",
            queryParams: ["foo": "bar", "baz": "qux"]
        )
        XCTAssertTrue(link.contains("foo=bar"))
        XCTAssertTrue(link.contains("baz=qux"))
    }

    func testGenerateDeepLinkWithoutQueryParams() {
        let handler = DeepLinkHandler()
        let link = handler.generateDeepLink(walletId: "metamask", uri: "wc:abc", queryParams: nil)
        XCTAssertEqual(link, "metamask:///wc?uri=wc:abc")
    }

    // MARK: - Universal Links

    func testGenerateUniversalLink() {
        let handler = DeepLinkHandler()
        let link = handler.generateUniversalLink(walletId: "metamask", uri: "wc:abc")
        XCTAssertNotNil(link)
        XCTAssertTrue(link?.hasPrefix("https://") ?? false)
        XCTAssertTrue(link?.contains("metamask.app.link") ?? false)
        XCTAssertTrue(link?.contains("wc:abc") ?? false)
    }

    func testUniversalLinkForWalletWithoutUniversalDomain() {
        let handler = DeepLinkHandler()
        // rabby has no universalDomain
        let link = handler.generateUniversalLink(walletId: "rabby", uri: "wc:test")
        XCTAssertNil(link)
    }
}
